File: sdk/python/dc1_provider/_http.py
"""Internal HTTP transport for the dc1_provider SDK.

Uses only stdlib (urllib) — no third-party dependencies required.
Provider auth uses the ``x-provider-key`` header for POST endpoints and
the ``?key=`` query param for GET endpoints (matching the backend contract).
"""
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Optional

from .exceptions import AuthError, DC1APIError

_SDK_VERSION = "0.1.0"


class _HttpClient:
    """Minimal stdlib-only HTTP client for the DC1 provider API."""

    def __init__(self, api_key: str, base_url: str, timeout: int):
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout

    # ------------------------------------------------------------------
    # Public helpers
    # ------------------------------------------------------------------

    def get(self, path: str, params: Optional[dict] = None) -> Any:
        """GET request.  Injects ``?key=`` auth param automatically."""
        p = dict(params or {})
        p["key"] = self._api_key
        return self._request("GET", path, params=p)

    def post(self, path: str, body: Optional[dict] = None, *, auth_header: bool = True) -> Any:
        """POST request.  Injects ``x-provider-key`` header when *auth_header* is True."""
        return self._request("POST", path, body=body, auth_header=auth_header)

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[dict] = None,
        params: Optional[dict] = None,
        auth_header: bool = True,
    ) -> Any:
        url = self._base_url + path
        if params:
            url = url + "?" + urllib.parse.urlencode(params)

        data = json.dumps(body).encode() if body is not None else None
        headers: dict[str, str] = {
            "Content-Type": "application/json",
            "User-Agent": f"dc1-provider-python/{_SDK_VERSION}",
        }
        if auth_header:
            headers["x-provider-key"] = self._api_key

        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                raw = resp.read().decode()
                return json.loads(raw) if raw else {}
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode()
            try:
                payload = json.loads(raw)
            except Exception:
                payload = {"error": raw or str(exc)}

            msg = payload.get("error", f"HTTP {exc.code}")
            if exc.code in (401, 403):
                raise AuthError(msg, status_code=exc.code, response=payload) from exc
            raise DC1APIError(msg, status_code=exc.code, response=payload) from exc
        except urllib.error.URLError as exc:
            raise DC1APIError(f"Connection error: {exc.reason}") from exc

File: sdk/node/src/http.ts
/**
 * Minimal HTTP client using Node.js built-in `https`/`http` (no third-party deps).
 */
import * as https from 'https';
import * as http from 'http';
import { URL } from 'url';
import { APIError, AuthError } from './errors';

export class HttpClient {
  /** Exposed so resources can inject it as a query param for endpoints that require ?key= */
  readonly apiKey: string;

  constructor(
    apiKey: string,
    private readonly baseUrl: string,
    private readonly timeoutMs: number,
  ) {
    this.apiKey = apiKey;
  }

  private request<T>(method: string, path: string, body?: unknown, params?: Record<string, string>): Promise<T> {
    return new Promise((resolve, reject) => {
      const url = new URL(this.baseUrl + path);
      if (params) {
        for (const [k, v] of Object.entries(params)) {
          url.searchParams.set(k, v);
        }
      }

      const payload = body != null ? JSON.stringify(body) : undefined;
      const options: http.RequestOptions = {
        method,
        hostname: url.hostname,
        port: url.port || (url.protocol === 'https:' ? 443 : 80),
        path: url.pathname + url.search,
        headers: {
          'Content-Type': 'application/json',
          'x-renter-key': this.apiKey,
          'User-Agent': 'dc1-renter-sdk/0.1.0',
          ...(payload ? { 'Content-Length': Buffer.byteLength(payload) } : {}),
        },
        timeout: this.timeoutMs,
      };

      const transport = url.protocol === 'https:' ? https : http;
      const req = transport.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (c: Buffer) => chunks.push(c));
        res.on('end', () => {
          const raw = Buffer.concat(chunks).toString();
          let parsed: unknown;
          try {
            parsed = raw ? JSON.parse(raw) : {};
          } catch {
            parsed = { error: raw };
          }

          const statusCode = res.statusCode ?? 0;
          if (statusCode >= 200 && statusCode < 300) {
            resolve(parsed as T);
          } else if (statusCode === 401) {
            reject(new AuthError((parsed as Record<string, string>)?.error));
          } else {
            const resp = (parsed ?? {}) as Record<string, unknown>;
            reject(new APIError(
              (resp.error as string) ?? `HTTP ${statusCode}`,
              statusCode,
              resp,
            ));
          }
        });
      });

      req.on('error', (e) => reject(new APIError(`Connection error: ${e.message}`)));
      req.on('timeout', () => {
        req.destroy();
        reject(new APIError(`Request timed out after ${this.timeoutMs}ms`));
      });

      if (payload) req.write(payload);
      req.end();
    });
  }

  get<T>(path: string, params?: Record<string, string>): Promise<T> {
    return this.request<T>('GET', path, undefined, params);
  }

  post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('POST', path, body);
  }
}
